### PR TITLE
Added 3 features + dependency upgrade

### PR DESCRIPTION
**Pull request by rhulshof**
- Added support for multiple traces
- Added support for creating an Excel tearsheet file
- Added config option for maximum exposure (to be able to spend i.e. 100% of balance into 1 single coin)
- Upgrade TA-Lib dependency to version 0.4.20 (from 0.4.19)

### CODE REVIEW

File: modules/stats/tradingmodule.py [hunk 1]
```diff
@@ -102,7 +103,7 @@ def open_trade(self, ohlcv: dict, data_dict: dict) -> None:
             return
 
         # Define spend amount based on realised profit
-        spend_amount = (1. / self.amount_of_pairs) * self.realised_profit
+        spend_amount = (self.max_exposure / 100.) * self.realised_profit
```
Comment (geordle) [inline modules/stats/tradingmodule.py:106]: Unfortunately we'll need to think about this a bit more before we can add this feature. In its current state it would significantly impact existing strategies. My thought is that the previous implementation was the way it is so that strategies on pairs would always have enough balance to get into position (this makes results more reproducible. Just think about the order in which trades are executed having effect on results in this case). Could you maybe explain the behaviour you're trying to achieve here? Maybe we can come up with a different solution together that fits your needs (exposure per pair?, dynamic stake amount?) cc: @DemianVoorhagen6 @sheridavandenbent

File: modules/output/plots.py [hunk 2]
```diff
@@ -17,8 +17,7 @@
 def plot_sizes(subplot_indicator, df):
     rows = 1
     for ind in subplot_indicator:
-        if ind in df.columns.values:
-            rows += 1
+        rows += 1
```
Comment (geordle) [inline modules/output/plots.py:20]: `rows = 1 + len(subplot_indicator)`

File: modules/output/__init__.py [hunk 3]
```diff
@@ -79,3 +86,11 @@ def log_trades(stats: TradingStats):
...
     with open('./data/backtesting-data/trades_log.json', 'w', encoding='utf-8') as f:
         f.write(trades_json)
+
+def create_tearsheet(stats: TradingStats):
+    dict_count = len(stats.trades)
+    df = pd.DataFrame(stats.trades[0].__dict__, index=[0])
+    for i in range(1,dict_count-1):
+        df = df.append(stats.trades[i].__dict__, ignore_index=True)
+
+    df.to_excel('data/backtesting-data/tearsheet.xlsx')
```
Comment (geordle) [inline modules/output/__init__.py:96]: I needed to install a dependency for this called `openpyxl`. Maybe this needs to be added to `requirements.txt`

**Review by geordle — COMMENTED** [review]
Awesome! Code is good & even follows our "undocumented" guidelines, kudos. There are some things regarding max_exposure that will take some deliberation but rest looks good.

File: modules/stats/tradingmodule.py (hunk 1 above)
Comment (rhulshof) [reply modules/stats/tradingmodule.py:106]: The goal is to analyze what the correlation is between several coins. I.e. one can decide to spend 100% of the balance on whichever coin displayes a buy signal first. Or one can decide to spend 50% per coin, whilst watching at 3 coins separately. So whichever 2 coins give their buy signal first, will end up with 50% of the balance. The backtester already supports this by displaying a "cannot buy" message in the terminal message, whenever the balance is already used on other coins.

Despite the above, I totally agree with your objections. It indeed would break existing setups. What I can do, is make sure the behavior remains as it was whenever the configuration setting is missing and/or None. Only when it actually has a percentage value, it will follow the above logic. This way, backtesters are completely in control. Do you agree? If yes, I will update the pull request.

Comment (geordle) [reply modules/stats/tradingmodule.py:106]: That `cannot buy` message should not be possible in recent versions (from 2 months ago onwards) of the engine.

I'm OK with that but, what if we changed the `max_exposure` implementation a little bit. Currently, when wanting to distribute balance across trading pairs you'd have to manually  `100 / amount_of_pairs` and fill that for the `max_exposure` variable. Everytime you change your pairs you'd have to update this variable. If we change this to `exposure_per_trade` (note the ~~max~~, since it will actually decide exposure, not cap it) we could spread funds evenly across pairs with a value of `1.`, flatten sways by lowering it `FI: .5` or achieve what you're trying to do by setting it by raising `2`.

the adjusted formula would be something in the lines of
```python
exposure_per_pair = config["exposure_per_trade"] or 1.
splits_amount = min(self.max_open_trades, self.amount_of_pairs) # recent change
spend_amount = ((1. / splits_amount) * exposure_per_pair) * self.realised_profit
```

What do you think?

Comment (geordle) [reply modules/stats/tradingmodule.py:106]: I'm not yet decided on whether we should use ratio or percentages, or on the naming

File: modules/output/__init__.py (hunk 3 above)
Comment (rhulshof) [reply modules/output/__init__.py:96]: Indeed. Will do.

File: modules/stats/tradingmodule.py (hunk 1 above)
Comment (rhulshof) [reply modules/stats/tradingmodule.py:106]: Stoploss setting already uses percentages, not ratio's. Therefore used the same here. Should be the way to go right?

Comment (rhulshof) [reply modules/stats/tradingmodule.py:106]: I like your change and will build it accordingly.

File: modules/output/plots.py (hunk 2 above)
Comment (rhulshof) [reply modules/output/plots.py:20]: Clever :)

**Comment by rhulshof** [comment]
Request was updated. It now includes all discussed topics.

File: modules/public/trading_stats.py [hunk 4]
```diff
@@ -11,6 +11,7 @@ class MainResults:
     tested_from: datetime
     tested_to: datetime
     max_open_trades: int
+    exposure_per_trade: int
```
Comment (geordle) [inline modules/public/trading_stats.py:14]:
```suggestion
    exposure_per_trade: float
```

File: modules/stats/stats_config.py [hunk 5]
```diff
@@ -10,10 +10,12 @@ class StatsConfig:
     stoploss: float
     stoploss_type: str
     max_open_trades: int
+    exposure_per_trade: int
```
Comment (geordle) [inline modules/stats/stats_config.py:13]:
```suggestion
    exposure_per_trade: float
```

File: modules/stats/tradingmodule_config.py [hunk 6]
```diff
@@ -8,6 +8,7 @@
 class TradingModuleConfig:
     fee: float
     max_open_trades: int
+    exposure_per_trade: int
```
Comment (geordle) [inline modules/stats/tradingmodule_config.py:11]:
```suggestion
    exposure_per_trade: float
```

**Review by geordle — APPROVED** [review]

File: config.json [hunk 7]
```diff
@@ -2,6 +2,7 @@
   "exchange": "binance",
   "timeframe": "5m",
   "max-open-trades": 3,
+  "exposure-per-trade": 1,
```
Comment (geordle) [inline config.json:5]:
```suggestion
  "exposure-per-trade": 100,
```

File: config.example.json [hunk 8]
```diff
@@ -2,6 +2,7 @@
   "exchange": "binance",
   "timeframe": "5m",
   "max-open-trades": "3",
+  "exposure-per-trade": "100",
```
Comment (geordle) [inline config.example.json:5]:
```suggestion
  "exposure-per-trade": 100,
```

File: modules/setup/config/__init__.py [hunk 9]
```diff
@@ -72,7 +73,9 @@ async def create(args):
         config_module.stoploss = config["stoploss"]
         config_module.stoploss_type = config["stoploss-type"]
         config_module.max_open_trades = config["max-open-trades"]
+        config_module.exposure_per_trade = config["exposure-per-trade"]
```
Comment (geordle) [inline modules/setup/config/__init__.py:76]:
```suggestion
        config_module.exposure_per_trade = config["exposure-per-trade"] / 100.
```

File: test/stats/stats_test_utils.py [hunk 10]
```diff
@@ -10,6 +10,7 @@
 StatsModuleFactory = [[], StatsModule]
 
 max_open_trades = 3
+exposure_per_trade = 100
```
Comment (geordle) [inline test/stats/stats_test_utils.py:13]:
```suggestion
exposure_per_trade = 1.
```

File: modules/output/results.py [hunk 11]
```diff
@@ -176,6 +175,7 @@ def create_settings_table(self: MainResults, currency_symbol, justification, tes
                            f"{round(self.starting_capital, 2)} {currency_symbol}")
     settings_table.add_row("Fee percentage", f"{self.fee} %")
     settings_table.add_row("Max. open trades", str(self.max_open_trades))
+    settings_table.add_row("Max. exposure per trade", str(self.exposure_per_trade))
```
Comment (geordle) [inline modules/output/results.py:178]:
```suggestion
    settings_table.add_row("Exposure per trade", str(self.exposure_per_trade * 100) + " %")
```

**Review by geordle — CHANGES_REQUESTED** [review]
There are a few things left regarding that exposure ratio format. think percentages would be nice for user input, and internally normalized (1.) ratios are nice because they"re normal :). i think tests should also succeed after these changes

File: test/stats/stats_test_utils.py [hunk 12]
```diff
@@ -23,6 +24,7 @@ class StatsFixture:
     def __init__(self, pairs: list):
         self.stats_config = StatsConfig(
             max_open_trades=max_open_trades,
+            exposure_per_trade=exposure_per_trade,
```
Comment (geordle) [inline test/stats/stats_test_utils.py:27]:
```suggestion
            exposure_per_trade=exposure_per_trade,
             tearsheet=False,
```

**Comment by rhulshof** [comment]
On my Windows machine all pytest's are failing. Even when running it on the "main" branch. All other issues I have fixed. Do you know why it's still failing?

**Comment by geordle** [comment]
> On my Windows machine all pytest's are failing. Even when running it on the "main" branch. All other issues I have fixed. Do you know why it's still failing?

Hmm, weird. In pipelines and local they seem to work for me. If end-to-end tests succeed I'm completely fine with the changes. Thanks a lot!

**Comment by geordle** [comment]
> On my Windows machine all pytest's are failing. Even when running it on the "main" branch. All other issues I have fixed. Do you know why it's still failing?

I'll take a look on windows as soon as I can

File: modules/setup/config/__init__.py [hunk 13]
```diff
@@ -72,7 +74,9 @@ async def create(args):
         config_module.stoploss = config["stoploss"]
         config_module.stoploss_type = config["stoploss-type"]
         config_module.max_open_trades = config["max-open-trades"]
+        config_module.exposure_per_trade = config["exposure-per-trade"] / 100.
         config_module.plots = config["plots"]
+        config_module.tearsheet = config["tearsheet"]
```
Comment (geordle) [inline modules/setup/config/__init__.py:79]: Ah yeah, this is a breaking change. existing configs will struggle on this (hence the failing end-2-end tests).
The following will access safely and provide a default value.
```suggestion
        config_module.exposure_per_trade = config.get("exposure-per-trade", 100.) / 100.
        config_module.plots = config["plots"]
        config_module.tearsheet = config.get("tearsheet", False)
```

**Review by geordle — CHANGES_REQUESTED** [review]

File: modules/setup/config/__init__.py (hunk 13 above)
Comment (geordle) [reply modules/setup/config/__init__.py:79]: I'm checking pytest on windows now but should stop us from merging

**Comment by geordle** [comment]
Tests on windows are failing because timestamp is broken for one day past epoch on windows. https://stackoverflow.com/a/67904074

Might fix this later

**Review by geordle — APPROVED** [review]
